handlers/proxy_handler_test.cc

```diff
@@ -0,0 +1,601 @@
+	std::string new_host;
+	handler->ParseRedirectLocation("www.ucla.edu", &new_path, &new_host);
+
+	EXPECT_EQ(new_host, "www.ucla.edu");
+	EXPECT_EQ(new_path, "/");
+}
+
+TEST(ProxyHandlerTest_ParseRedirectLocation, http) {
+	auto handler = MockSetupProxyHandler::createForTesting("/", "ucla.edu", "80");
+
+	std::string new_path;
+	std::string new_host;
+	handler->ParseRedirectLocation("http://www.ucla.edu/some/path", &new_path, &new_host);
+
+	EXPECT_EQ(new_host, "www.ucla.edu");
+	EXPECT_EQ(new_path, "/some/path");
+}
+
+TEST(ProxyHandlerTest_ParseRedirectLocation, https) {
+	auto handler = MockSetupProxyHandler::createForTesting("/", "ucla.edu", "80");
+
+	std::string new_path;
+	std::string new_host;
+	handler->ParseRedirectLocation("https://www.ucla.edu/some/path", &new_path, &new_host);
+
+	EXPECT_EQ(new_host, "www.ucla.edu");
+	EXPECT_EQ(new_path, "/some/path");
+}
+
+
+TEST(ProxyHandlerTest_ModifyRequestForProxy, simple) {
+	auto handler = MockSetupProxyHandler::createForTesting("/", "ucla.edu", "80");
+
+	string reqStr = (
+			"GET / HTTP/1.1\r\n"
+			"User-Agent: Mozilla/1.0\r\n"
+			"\r\n"
+			);
+
+	std::unique_ptr<Request> request = Request::Parse(reqStr);
+	MutableRequest modified_request(*request);
+	handler->ModifyRequestForProxy(*request, &modified_request);
+
+	string result = modified_request.ToString();
+
+	EXPECT_THAT(result, HasSubstr("GET / HTTP/1.1\r\n"));
+	EXPECT_THAT(result, HasSubstr("User-Agent: Mozilla/1.0\r\n"));
+	EXPECT_THAT(result, HasSubstr("Connection: close\r\n"));
+	EXPECT_THAT(result, HasSubstr("Host: ucla.edu"));
+	EXPECT_THAT(result, HasSubstr("\r\n\r\n"));
+}
+
+TEST(ProxyHandlerTest_ModifyRequestForProxy, overwriteHeaders) {
+	auto handler = MockSetupProxyHandler::createForTesting("/", "ucla.edu", "80");
+
+	string reqStr = (
+			"GET / HTTP/1.1\r\n"
+			"Connection: keep-alive\r\n"
+			"\r\n"
+			);
+
+	std::unique_ptr<Request> request = Request::Parse(reqStr);
+	MutableRequest modified_request(*request);
+	handler->ModifyRequestForProxy(*request, &modified_request);
+
+	string result = modified_request.ToString();
+
+	EXPECT_THAT(result, HasSubstr("GET / HTTP/1.1\r\n"));
+	EXPECT_THAT(result, HasSubstr("Connection: close\r\n"));
+	EXPECT_THAT(result, HasSubstr("\r\n\r\n"));
+}
+
+TEST(ProxyHandlerTest_ModifyRequestForProxy, RemovePrefix) {
+	auto handler = MockSetupProxyHandler::createForTesting("/proxy", "ucla.edu", "80");
+
+	string reqStr = (
+			"GET /proxy/some/path HTTP/1.1\r\n"
+			"User-Agent: Mozilla/1.0\r\n"
+			"\r\n"
+			);
+
+	std::unique_ptr<Request> request = Request::Parse(reqStr);
+	MutableRequest modified_request(*request);
+	handler->ModifyRequestForProxy(*request, &modified_request);
+
+	string result = modified_request.ToString();
+
+	EXPECT_THAT(result, HasSubstr("GET /some/path HTTP/1.1\r\n"));
+	EXPECT_THAT(result, HasSubstr("User-Agent: Mozilla/1.0\r\n"));
+	EXPECT_THAT(result, HasSubstr("Connection: close\r\n"));
+	EXPECT_THAT(result, HasSubstr("\r\n\r\n"));
+}
+
```

Comment:
Currently, you can only set the proxy host to be a path-less host. For example, you'd configure the proxy to have the host `ucla.edu`. If you tried to use `ucla.edu/static`, it would fail to work as a proxy. If you configure the path of the ProxyHandler to `/proxy`, then anything after that will be taken as the new URI. So for example, if your config is
```
port 2020;
path /proxy ProxyHandler {
   host: ucla.edu;
}
```
then a request to `localhost:2020/proxy/something` would request `ucla.edu/something`.
I do have a test for this situation. By design, you can't have the proxy route requests onto a different path base on the remote server. Maybe that could be a future feature? It's a little out of spec for this particular project I think?